### PR TITLE
[TIR] Shuffle in PointerValueTypeRewrite for scalar reads 

### PR DESCRIPTION
**Pull request by vinx13**
Added an option `rewrite_scalar_read_to_vector_shuffle` in `PointerValueTypeRewrite` (currently
only enabled for Vulkan). When enabled, when a buffer has both scalar
and vector reads, the buffer will be vectorized if possible and scalar
reads will be achieved via T.Shuffle.

Close https://github.com/apache/tvm/issues/15463.

cc @tqchen @Lunderberg @sunggg

### CODE REVIEW

**Comment by tqchen** [comment]
@Lunderberg do you mind help reviewing this

**Comment by Lunderberg** [comment]
I can, though I probably won't have time available to do so until later next week.

**Comment by tqchen** [comment]
ah ok, i can help to take a look  then

File: src/tir/transforms/storage_rewrite.cc [hunk 1]
```diff
@@ -1304,6 +1307,18 @@ class VectorTypeAccessChecker : public StmtExprVisitor {
...
     }
 
+    if (detect_scalar_read_patterns_ && is_buffer_load && indices.size()) {
+      const PrimExpr last_dim_index = indices[indices.size() - 1];
+      if (last_dim_index.dtype().lanes() == 1) {
+        arith::ModularSet me = analyzer_.modular_set(last_dim_index);
+        if (me->coeff > 0) {
+          // When coeff == 0, the index is constant and doesn't need to be recorded since it can
+          // always be rewritten to shuffle.
+          var_info.access_dtype.insert(access_dtype.with_lanes(me->coeff));
```
Comment (tqchen) [inline src/tir/transforms/storage_rewrite.cc:1317]: i think we need a data structure that captures two things: 

```c++
struct VarReadInfo {
   DataType access_dtype;
   // maintained as GCD of all coef of access index
   int64_t simd_coeff;
};
```
This is mainly to ensure that we don't end up use a very large vector here

**Review by tqchen — COMMENTED** [review]
Thanks wuwei, some comment

File: src/tir/transforms/storage_rewrite.cc (hunk 1 above)
Comment (vinx13) [reply src/tir/transforms/storage_rewrite.cc:1317]: access_dtype also contains dtype for write, so the result dtype will be eventually bound by the vectorization lanes of writes

File: src/tir/transforms/storage_rewrite.cc [hunk 2]
```diff
@@ -1083,17 +1084,14 @@ struct BufferVarInfo {
...
-    // buffer, and if that access size is compatible with the array
-    // size, then the buffer is vectorizable.  In the future, this
-    // could be improved to allow vectorized buffer access of size
-    // GCD(*lanes_used), if necessary.
     int preferred_lanes = element_dtype.lanes();
-    if ((element_dtype.lanes() == 1) && (access_dtype.size() == 1)) {
+    if (element_dtype.lanes() == 1) {
+      int lanes = access_dtype.begin()->lanes();
+      for (auto dtype : access_dtype) {
+        lanes = std::gcd(lanes, dtype.lanes());
```
Comment (tqchen) [inline src/tir/transforms/storage_rewrite.cc:1091]: use https://github.com/apache/tvm/blob/main/src/arith/int_operator.h#L171

File: src/tir/transforms/storage_rewrite.cc (hunk 1 above)
Comment (tqchen) [reply src/tir/transforms/storage_rewrite.cc:1317]: the main corner case would be when we do not have writes and only reads.

Comment (vinx13) [reply src/tir/transforms/storage_rewrite.cc]: updated

**Review by tqchen — APPROVED** [review]

File: src/tir/transforms/storage_rewrite.cc [hunk 3]
```diff
@@ -1081,19 +1088,24 @@ struct BufferVarInfo {
...
+    if (access_dtype.empty()) {
+      return element_dtype;
+    }
     DataType preferred_base_type = *base_access_dtype.begin();
 
-    // If there is only one vectorizable size used to access the
-    // buffer, and if that access size is compatible with the array
-    // size, then the buffer is vectorizable.  In the future, this
-    // could be improved to allow vectorized buffer access of size
-    // GCD(*lanes_used), if necessary.
```
Comment (vinx13) [inline src/tir/transforms/storage_rewrite.cc:1090]: it seems taking gcd can cause regressions in some source codegen cases (e.g cuda). For example, when a buffer is accessed via `ramp(lanes=2)` and `ramp(lanes=4`), rewriting the buffer to float2 will disable the vectorization for float4.

Comment (tqchen) [reply src/tir/transforms/storage_rewrite.cc:1090]: How about we only trigger rewriting if:
- there is vector write
- original dtype was float4
- there is only scalar read access

In a sense that comparing to original case, the only difference is that we allow scalar read from vector write

Comment (vinx13) [reply src/tir/transforms/storage_rewrite.cc:1090]: That makes sense, I've updated rewriting. Now the rewriting only trigger if there is one vector access dtype with lanes same as the coeff scalar access.

**Comment by MasterJH5574** [comment]
Hi @vinx13, we noticed that this PR breaks the WebGPU codegen as the WebGPU codegen right now does not support tir::ShuffleNode. Therefore, exceptions are thrown in pass `PointerValueTypeRewrite` at the beginning of WebGPU codegen.

Here is a reproducible example https://gist.github.com/MasterJH5574/3e6141a1d6dcafd383ed6f7ac27e3318. Need to run it under this commit https://github.com/mlc-ai/relax/commit/631f37b6bf8b101d16ecc55de7e6a749a3588570 since the WebGPU codegen in mlc-ai/relax is slightly different from the codegen here due to some conflicts.

cc @tqchen

**Comment by vinx13** [comment]
Is it possible to support it in codegen? Usually this can be supported via element extraction  e.g `vex.x/y/z`. Alternatively we can set `rewrite_scalar_access_to_vector_shuffle` to false

**Comment by tqchen** [comment]
I think we should support via codegen

**Comment by CharlieFRuan** [comment]
Will look into adding Shuffle support for WebGPU!

**Comment by guoyaol** [comment]
Hi @vinx13 
I was building TVM mod for stable diffusion, I found some kernels failed to pass the ICHECK in this PR. Wonder if you have any insights about how to tackle this? Thanks.
```
  File "/Users/guoyaoli/tvm_work/tvm/src/tir/transforms/storage_rewrite.cc", line 1466
InternalError: Check failed: (me->coeff == 0 || info.factor() % me->coeff == 0) is false: 
```
here's a minimum reproducible script, should reproduce the same error running on up-to-date unity branch: https://gist.github.com/guoyaol/af7d1161124987b69b8eb2744b1c399e

cc @MasterJH5574 @tqchen

**Comment by vinx13** [comment]
@guoyaol it may be unaligned vectorizations or a false alarm when the arithmetic analyzer can't handle the index pattern

**Pull request merged**